From f64967aaa784f0c9bebf0f58ac3b239a093dee00 Mon Sep 17 00:00:00 2001
From: Vladimir Barinov <vladimir.barinov@cogentembedded.com>
Date: Fri, 19 Jul 2019 14:05:43 +0300
Subject: [PATCH] media: i2c: ar0233: fix artifact line at vflip

This fixes artifact line in super-exposure mode at vflip=1

Signed-off-by: Vladimir Barinov <vladimir.barinov@cogentembedded.com>
---
 drivers/media/i2c/soc_camera/ar0231.c      | 2 +-
 drivers/media/i2c/soc_camera/ar0233.c      | 6 ++++++
 drivers/media/i2c/soc_camera/ar0233_rev2.h | 1 +
 3 files changed, 8 insertions(+), 1 deletion(-)

diff --git a/drivers/media/i2c/soc_camera/ar0231.c b/drivers/media/i2c/soc_camera/ar0231.c
index b239f56..dfca9e2 100644
--- a/drivers/media/i2c/soc_camera/ar0231.c
+++ b/drivers/media/i2c/soc_camera/ar0231.c
@@ -93,7 +93,7 @@ static int ar0231_set_window(struct v4l2_subdev *sd)
 	struct i2c_client *client = v4l2_get_subdevdata(sd);
 	struct ar0231_priv *priv = to_ar0231(client);
 
-	dev_err(&client->dev, "L=%d T=%d %dx%d\n", priv->rect.left, priv->rect.top, priv->rect.width, priv->rect.height);
+	dev_dbg(&client->dev, "L=%d T=%d %dx%d\n", priv->rect.left, priv->rect.top, priv->rect.width, priv->rect.height);
 
 	/* horiz crop start */
 	reg16_write16(client, 0x3004, priv->rect.left + AR0231_X_START);
diff --git a/drivers/media/i2c/soc_camera/ar0233.c b/drivers/media/i2c/soc_camera/ar0233.c
index 1b7ba22..71bff3f 100644
--- a/drivers/media/i2c/soc_camera/ar0233.c
+++ b/drivers/media/i2c/soc_camera/ar0233.c
@@ -328,6 +328,12 @@ static int ar0233_s_ctrl(struct v4l2_ctrl *ctrl)
 		else
 			val &= ~(1 << 15);
 		ret |= reg16_write16(client, 0x3040, val);
+		ret = reg16_read16(client, 0x350e, &val);
+		if (ctrl->val)
+			val |= (1 << 0);
+		else
+			val &= ~(1 << 0);
+		ret |= reg16_write16(client, 0x350e, val);
 		break;
 	case V4L2_CID_MIN_BUFFERS_FOR_CAPTURE:
 		ret = 0;
diff --git a/drivers/media/i2c/soc_camera/ar0233_rev2.h b/drivers/media/i2c/soc_camera/ar0233_rev2.h
index 66c79ca..7085fbc 100644
--- a/drivers/media/i2c/soc_camera/ar0233_rev2.h
+++ b/drivers/media/i2c/soc_camera/ar0233_rev2.h
@@ -2266,6 +2266,7 @@ static const struct ar0233_reg ar0233_rev2_O1_Recommended_Defaults_SE_T1_LIN_T2[
 
 static const struct ar0233_reg ar0233_rev2_disable_embed_data_stat[] = {
 {0x3040, 0xC000}, //Embedded stat2 and data2 rows, hflip/vflip=1
+{0x350e, 0x2089}, // bit0 must be set for vflip=1
 #ifdef AR0233_EMBEDDED_LINE
 {0x3064, 0x0180}, // SMIA_TEST: enable emb data and stats
 #else
-- 
2.7.4

